retrieval/discovery/kubernetes.go

```diff
@@ -0,0 +1,593 @@
+				log.Errorf("Unable to watch services: %s", err)
+				return
+			}
+			kd.servicesResourceVersion = event.Service.ObjectMeta.ResourceVersion
+			events <- &event
+		}
+	}, retryDelay, kd.runDone)
+}
+
+func (kd *KubernetesDiscovery) updateService(service *kubernetes.Service, eventType kubernetes.EventType, ch chan<- *config.TargetGroup) {
+	kd.servicesMu.Lock()
+	defer kd.servicesMu.Unlock()
+	name := service.ObjectMeta.Name
+	namespace := service.ObjectMeta.Namespace
+	scrape := kd.checkScrape(service)
+	_, ok := kd.services[namespace][name]
+	switch eventType {
+	case kubernetes.Deleted:
+		if ok {
+			kd.deleteService(service, ch)
+		}
+	case kubernetes.Modified:
+		// Exists & no scrape - delete.
+		if ok && !scrape {
+			kd.deleteService(service, ch)
+			return
+		}
+
+		if scrape {
+			// Must be scrapable so let's add it to ensure labels are updated properly at least.
+			kd.addService(service, ch)
+		}
+	case kubernetes.Added:
+		if scrape {
+			kd.addService(service, ch)
+		}
+	}
+}
+
+func (kd *KubernetesDiscovery) deleteService(service *kubernetes.Service, ch chan<- *config.TargetGroup) {
+	tg := &config.TargetGroup{Source: kubernetesSourceServicePrefix + ":" + service.ObjectMeta.Namespace + "/" + service.ObjectMeta.Name}
+	ch <- tg
+	delete(kd.services[service.ObjectMeta.Namespace], service.ObjectMeta.Name)
+	if len(kd.services[service.ObjectMeta.Namespace]) == 0 {
+		delete(kd.services, service.ObjectMeta.Namespace)
+	}
+}
+
+func (kd *KubernetesDiscovery) addService(service *kubernetes.Service, ch chan<- *config.TargetGroup) {
+	namespace, ok := kd.services[service.ObjectMeta.Namespace]
+	if !ok {
+		namespace = make(map[string]*kubernetes.Service)
+		kd.services[service.ObjectMeta.Namespace] = namespace
+	}
+	namespace[service.ObjectMeta.Name] = service
+	endpointURL := fmt.Sprintf(kubernetesServiceEndpointsURL, service.ObjectMeta.Namespace, service.ObjectMeta.Name)
+	res, err := kd.client.Get(kd.conf.Server + endpointURL)
+	if err != nil {
+		log.Errorf("Error getting service endpoints: %s", err)
+		return
+	}
+	if res.StatusCode != http.StatusOK {
+		log.Errorf("Failed to get service endpoints: %s", res.StatusCode)
+		return
+	}
+
+	var endpoints kubernetes.Endpoints
+	if err := json.NewDecoder(res.Body).Decode(&endpoints); err != nil {
+		log.Errorf("Error getting service endpoints: %s", err)
+		return
+	}
+
+	kd.updateServiceTargetGroup(service, &endpoints, ch)
+}
+
+func (kd *KubernetesDiscovery) updateServiceTargetGroup(service *kubernetes.Service, endpoints *kubernetes.Endpoints, ch chan<- *config.TargetGroup) {
+	tg := &config.TargetGroup{
+		Source: kubernetesSourceServicePrefix + ":" + service.ObjectMeta.Namespace + "/" + service.ObjectMeta.Name,
+		Labels: clientmodel.LabelSet{
+			kubernetesServiceNamespaceLabel: clientmodel.LabelValue(service.ObjectMeta.Namespace),
+			kubernetesServiceNameLabel:      clientmodel.LabelValue(service.ObjectMeta.Name),
+		},
+	}
+
+	for k, v := range service.ObjectMeta.Labels {
```

Comment:
> All I see on that PR are moving of such labels into meta labels, and removal of deployment-specific code. #979 filed for removal of remaining deployment-specific code.

The discussion was with @fabxc on an outdated diff - https://github.com/prometheus/prometheus/pull/902#discussion_r36508629. Happy to discuss this again though, but the same argument applies.

How about a label whitelist that could be a slice of regexes? All non-matching would go to meta labels, otherwise "proper" labels? That way users could use `.*` to pass through all labels. Default to making all meta labels? I would almost definitely be using this to whitelist everything of course :)

> We need to design for the most general case, and there being a single label called environment is not the most general case.

I'm not sure I follow. What do you mean by general case? IMO in this case users are not going to want to label everything separately, relabel extensively & correlate those labels to resources somehow. Reusing Kubernetes labels seems to be the most general case I can think of. This is a (potentially multi-tenant) environment where Prometheus can be provided as a drop-in service with very little configuration. To get this being used & gather further community feedback to improve I would suggest going with this 1-1 label mapping to begin with.